src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/HandlerProvider.cs

```diff
@@ -100,23 +108,29 @@ public ImmutableArray<RequestHandlerMetadata> GetRegisteredMethods()
                 var method = GetRequestHandlerMethod(handlerType, requestResponseType.RequestType, requestResponseType.RequestContext, requestResponseType.ResponseType);
                 CheckForDuplicates(method, methodHash);
 
-                requestHandlerDictionary.Add(new RequestHandlerMetadata(method, requestResponseType.RequestType, requestResponseType.ResponseType), new Lazy<IMethodHandler>(() => handler));
+                var requestHandlerMetadata = new RequestHandlerMetadata(method.Name, requestResponseType.RequestType, requestResponseType.ResponseType);
+                if (!requestHandlerDictionary.TryGetValue(requestHandlerMetadata, out var methodHandlers))
+                {
+                    methodHandlers = new List<Lazy<IMethodHandler, string?>>();
+                    requestHandlerDictionary.Add(requestHandlerMetadata, methodHandlers);
+                }
+                methodHandlers.Add(new Lazy<IMethodHandler, string?>(() => handler, method.Language));
             }
         }
 
         VerifyHandlers(requestHandlerDictionary.Keys);
 
         return requestHandlerDictionary.ToImmutable();
 
-        static void CheckForDuplicates(string methodName, HashSet<string> existingMethods)
+        static void CheckForDuplicates((string, string?) method, HashSet<(string, string?)> existingMethods)
         {
-            if (!existingMethods.Add(methodName))
+            if (!existingMethods.Add(method))
             {
-                throw new InvalidOperationException($"Method {methodName} was implemented more than once.");
+                throw new InvalidOperationException($"Method {method.Item1} was implemented more than once.");
             }
         }
 
-        static string GetRequestHandlerMethod(Type handlerType, Type? requestType, Type contextType, Type? responseType)
+        static (string Name, string? Language) GetRequestHandlerMethod(Type handlerType, Type? requestType, Type contextType, Type? responseType)
```

Comment:
lowercase names for tuples.  Or define a record struct for this concept.

Reply:
done